test user gets created in test database

natevenn opened the pull request:
add jasmine-node && request
add test script to package.json
create spec/routes/signup
create routes/api/v1/signup
config knexfile to have a test database
create database photo-test

#### Issue
Cannot get test data to get saved to test-db. 
Look in spec/routes/signup.spec.js
before action gets test-db and then migrates latest. Then sets test-db to app. Test db migrates fine, but when I make a post with data - data still gets sent to dev-db.

natevenn committed:
test user gets created in test database

add jasmine-node request
create spec/routes/signup
create routes/api/v1/signup
config knexfile to have a test database

natevenn commented on spec/routes/signup.spec.js:
```diff
@@ -0,0 +1,37 @@
+describe('api/v1/signin', function() {
+
+  beforeEach( (done) => {
+    var db = app.get('db');
```
This is where I thought I would be setting the test-db for making posts. the migrations to the test-db work but it doesnt set the test-db to app. I am not sure what I am missing here.

wilmoore commented on db/knex.js:
```diff
@@ -1,3 +1,4 @@
 var env = process.env.NODE_ENV || 'development';
 var config = require('../knexfile.js')[env];
+console.log('db config', config)
```
Investigate and use the `debug` package here instead of `console.log`.

wilmoore replied:
Investigate and use the `debug` package here instead of `console.log`.

wilmoore commented on package.json:
```diff
@@ -3,7 +3,8 @@
   "scripts": {
-    "start": "node ./bin/www"
+    "start": "node ./bin/www",
+    "test": "jasmine-node spec/ --captureExceptions --forceexit"
```
I would recommend that you use one of the following instead of `jasmine-node`:
- mocha
- ava
- lab
- tape

These are all much nicer testing frameworks compared to `jasmine` in my opinion.

natevenn replied:
Cool. I have now implemented mocha, since I have used it once before. Good to know that it is superior to jasmine.

wilmoore commented on routes/api/v1/signup.js:
```diff
@@ -0,0 +1,18 @@
+router.post('/', (req, res, next) => {
+  var user = req.body
+  var db = req.app.get('db')
+  if(user.username !== undefined && user.password !== undefined) {
```
I would suggest dropping the notion of passing a `username` and `password` for each request and instead use a JWT.

If you are not familiar with JWTs check out: https://youtu.be/X7t2pdJYHNI

natevenn replied:
I am really glad you brought this up. I came across JWT's when researching how to create a session. Thanks for the video. I will watch it and get a working implementation of JWTs.

wilmoore commented on package.json:
```diff
@@ -3,7 +3,8 @@
   "scripts": {
-    "start": "node ./bin/www"
+    "start": "node ./bin/www",
+    "test": "jasmine-node spec/ --captureExceptions --forceexit"
```
Add to the beginning of this script:

```
NODE_ENV=test 
```

natevenn replied:
Add!

wilmoore commented on spec/routes/signup.spec.js:
```diff
@@ -0,0 +1,37 @@
+let request = require('request');
+let app = require('../../app.js');
+
+process.env.NODE_ENV = 'test';
```
You can drop this since we've added it to the script which runs the tests.

wilmoore commented on spec/routes/signup.spec.js:
```diff
@@ -0,0 +1,37 @@
+      }
+    }
+
+    request.post(options, function(err, res, body) {
```
Instead of `request` I highly recommend using `supertest` or something like it. It will eliminate a lot of the boilerplate.

natevenn replied:
supertest is now my request.

natevenn committed:
add routes/index.js

wilmoore commented:
> This is where I thought I would be setting the test-db for making posts. the migrations to the test-db work but it doesn't set the test-db to app. I am not sure what I am missing here.

I don't necessarily think this is wrong; however, I do think this can be streamlined a bit by using an actual migrations package. Once you have that, you will mitigate the possibility of lots of duplication in your tests.

wilmoore reviewed (changes requested):
First time using the new code review feature in github. I believe this will not submit all of my comments as a single review.

natevenn committed:
add mocha chai and supertest to package.json

natevenn committed:
Merge branch 'master' of github.com:natevenn/photo-app-backend into test-user-saves

natevenn committed:
debug config in db/knex.js

natevenn commented on db/knex.js:
```diff
@@ -1,3 +1,7 @@
 var env = process.env.NODE_ENV || 'development';
 var config = require('../knexfile.js')[env];
+debug(config);
 module.exports = require('knex')(config);
```
@wilmoore 
So after reading the docs about the node debug package I was still pretty confused on how to implement the debugger. I can get the config when I run $ _DEBUG=app:config npm test_ and it shows that it is set to the test-db

_app:config { client: 'pg', connection: 'postgres://localhost/photo_test', pool: { min: 2, max: 10 } } +0ms_

but I do not how to investigate.  I do I get it to freeze?

wilmoore replied:
The `debug` package is a debug logger rather than an interactive debugger. If you want that, you'll want to use `node-inspector`. That being said, I pretty much never use `node-inspector` as `debug` is generally all I need.

wilmoore commented on test/signin.js:
```diff
@@ -0,0 +1,30 @@
+let chai = require('chai');
+let request = require('supertest')
+let should = chai.should();
+request = request('http://localhost:3000')
```
While this works, It's actually really bad practice to overwrite a reference with a reference of the same name.

I would recommend just doing the following if you don't want an extra reference:

```
let request = require('supertest')('http://localhost:3000');
```

wilmoore commented on test/signin.js:
```diff
@@ -0,0 +1,30 @@
+let chai = require('chai');
+let request = require('supertest')
+let should = chai.should();
+request = request('http://localhost:3000')
```
Actually, just thought about it, you should be doing something like:

```
request(app)
.post(...)
```

So you don't really need the `request = request('http://localhost:3000')`

natevenn replied:
Yes. request(app).post(..) worked!  thanks for catching that. Good to know that reassigning variables is not good practice. I will remember that.

natevenn committed:
test user signup route

add mocha, chai, and supertest
test duplicate users created returns msg "already exisits"
test 400 err when user submits with missing field
test user is successfully created

natevenn committed:
add comment in signup route post to hash password before save

natevenn commented on routes/api/v1/signup.js:
```diff
@@ -0,0 +1,29 @@
+        next(err);
+      }
+    }
+  });
```
@wilmoore 
This is my post for creating a user. I have written three tests (see below) and they are passing. My next thought would be to refactor. What are your thoughts on where this logic should live?

natevenn committed:
refactor signup post

natevenn committed:
delete commented out code

natevenn commented on routes/api/v1/signup.js:
```diff
@@ -0,0 +1,30 @@
+      var err = new Error('Not found');
+      err.status = 400;
+      next(err);
+    }
```
@wilmoore 
so what would be your recommendation for separating out the logic in the routes? Meaning where should it live?

wilmoore replied:
I'd start with a `lib` directory and go from there.

natevenn replied:
@wilmoore 
Now if its logic dealing with models (CRUD) I was thinking of a models folder like models/user.js and inside there have my CRUD functionality. But if it is logic that can be shared with all models you are saying to put that in a lib folder? Or are you saying inside lib put my models folder, similar to Rails app folder? sorry for the confusion. I have been looking for examples of how node folders are structured. Seems like everyone has their own opinion on the matter. Thanks for your input!

wilmoore replied:
Yes, it is basically what you described in that you definitely want a `lib` folder and inside there, you'd also have a `model` or `models` folder. I've seen it both ways with and without `models` being nested within `lib`. It's really a personal preference as there is no real right or wrong there as it relates to the resulting functionality.

I have gone back and forth myself about the most optimal and obvious directory structure and I've not yet settled on anything that works for every project. For some projects one structure seems to fit a little more than other approaches. 

At the end of the day, the rails app folder structure is fine and if that feels natural, go with it and you can change it down the line if you really feel strongly for some other pattern.

natevenn replied:
@wilmoore 
Thats helpful. I really dont have a preference at this point in my career. I tend to follow and try different approaches. It makes sense that it is personal preference and that it can change depending on the project. Im so use to Rails setting that structure for me that now within a different env Im, for the first time, having to think about these things. Thanks for your input. And happy holidays!

wilmoore replied:
Yes, Rails is definitely great in that it makes a lot of decisions for you so you don't have to; however, that doesn't really help you when you need to do something outside of Rails. It's not really the fault of Rails, it's their job to do that, so as a developer, you simply work through these things as they come up. We've all been there.

Happy holidays to you as well.
